kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/metadata.kt

```diff
@@ -123,3 +188,345 @@ internal fun Type.asTypeName(
+    }
+    if (supertype.element.kotlinMetadata == null) {
+      messager.printMessage(ERROR,
+          "@JsonClass can't be applied to $element: supertype $supertype is not a Kotlin type",
+          element)
+      return null
+    }
+    val supertypeProperties = declaredProperties(
+        supertype.element, supertype.resolver, constructor)
+    for ((name, property) in supertypeProperties) {
+      properties.putIfAbsent(name, property)
+    }
+  }
+  return TargetType(
+      typeName = element.asType().asTypeName(),
+      constructor = constructor,
+      properties = properties,
+      typeVariables = typeVariables,
+      isDataClass = proto.isDataClass,
+      visibility = proto.visibility.asKModifier())
+}
+
+/** Returns the properties declared by `typeElement`. */
+private fun declaredProperties(
+    typeElement: TypeElement,
+    typeResolver: TypeResolver,
+    constructor: TargetConstructor
+): Map<String, TargetProperty> {
+  val typeMetadata: KotlinClassMetadata = typeElement.kotlinMetadata as KotlinClassMetadata
+  val nameResolver = typeMetadata.data.nameResolver
+  val classProto = typeMetadata.data.classProto
+
+  val annotationHolders = mutableMapOf<String, ExecutableElement>()
+  val fields = mutableMapOf<String, VariableElement>()
+  val setters = mutableMapOf<String, ExecutableElement>()
+  val getters = mutableMapOf<String, ExecutableElement>()
+  for (element in typeElement.enclosedElements) {
+    if (element is VariableElement) {
+      fields[element.name] = element
+    } else if (element is ExecutableElement) {
+      when {
+        element.name.startsWith("get") -> {
+          val name = element.name.substring("get".length).decapitalizeAsciiOnly()
+          getters[name] = element
+        }
+        element.name.startsWith("is") -> {
+          val name = element.name.substring("is".length).decapitalizeAsciiOnly()
+          getters[name] = element
+        }
+        element.name.startsWith("set") -> {
+          val name = element.name.substring("set".length).decapitalizeAsciiOnly()
+          setters[name] = element
+        }
+      }
+
+      val propertyProto = typeMetadata.data.getPropertyOrNull(element)
+      if (propertyProto != null) {
+        val name = nameResolver.getString(propertyProto.name)
+        annotationHolders[name] = element
+      }
+    }
+  }
+
+  val result = mutableMapOf<String, TargetProperty>()
+  for (property in classProto.propertyList) {
+    val name = nameResolver.getString(property.name)
+    val type = typeResolver.resolve(property.returnType.asTypeName(
+        nameResolver, classProto::getTypeParameter, false))
+    val parameter = constructor.parameters[name]
+    val fieldElement = fields[name]
+    val annotationHolder = annotationHolders[name]
+    // Used for setter/getter/is lookups. Guaranteed to be safe because kotlin doesn't allow you to
+    // have both "AAA" and "aAA".
+    val decapitalizedName = name.decapitalizeAsciiOnly()
```

Comment:
I actually don't know how master passes without this currently

![image](https://user-images.githubusercontent.com/1361086/53314198-1bd40480-388b-11e9-84aa-829d6bb7bad7.png)